pkg/files/file_helpers.go

```diff
@@ -139,3 +139,62 @@ func CompareFileHash(fileOverview *mpi.FileOverview) (fileDiff map[string]*mpi.F
 
 	return fileDiff, fileContents, nil
 }
+
+// DetermineFileAction compares two lists of files, determines which file action is needed for each file,
+// then returns a list of files with the action set for each file
+func DetermineFileAction(currentFiles, modifiedFiles []*mpi.File) []*mpi.File {
+	// Go doesn't allow address of numeric constant
+	addAction := mpi.File_FILE_ACTION_ADD
+	updateAction := mpi.File_FILE_ACTION_UPDATE
+	deleteAction := mpi.File_FILE_ACTION_DELETE
+	unchangedAction := mpi.File_FILE_ACTION_UNCHANGED
+
+	filesWithActions := make([]*mpi.File, 0, len(modifiedFiles))
+	currentFilesMap := convertToMapOfFiles(currentFiles)
+	modifiedFilesMap := convertToMapOfFiles(modifiedFiles)
+
+	// if file is in currentFiles but not in modified files, file has been deleted
+	for _, currentFile := range currentFilesMap {
+		_, ok := modifiedFilesMap[currentFile.GetFileMeta().GetName()]
+		if !ok {
+			currentFile.Action = &deleteAction
+			filesWithActions = append(filesWithActions, currentFile)
+			delete(modifiedFilesMap, currentFile.GetFileMeta().GetName())
+
+			continue
+		}
+	}
+
+	for _, file := range modifiedFilesMap {
+		_, ok := currentFilesMap[file.GetFileMeta().GetName()]
+		// if file doesn't exist in the current files, file has been added
+		if !ok {
+			file.Action = &addAction
+			filesWithActions = append(filesWithActions, file)
+
+			continue
+			// if file currently exists and file hash is different, file has been updated
+		} else if file.GetFileMeta().GetHash() != currentFilesMap[file.GetFileMeta().GetName()].
+			GetFileMeta().GetHash() {
+			file.Action = &updateAction
+			filesWithActions = append(filesWithActions, file)
+
+			continue
+		}
+		// if file exists and file hash matches, file is unchanged
+		file.Action = &unchangedAction
+		filesWithActions = append(filesWithActions, file)
+	}
```

Comment:
This could be simplified somewhat to improve readabiility.
```suggestion
		// default to unchanged action.
		file.Action = &unchangedAction

		// if file doesn't exist in the current files, file has been added
		if !ok {
			file.Action = &addAction
		}
		
		// if file currently exists and file hash is different, file has been updated
		if ok && file.GetFileMeta().GetHash() != currentFilesMap[file.GetFileMeta().GetName()].
			GetFileMeta().GetHash() {
			file.Action = &updateAction
		}
		
		filesWithActions = append(filesWithActions, file)
	}
```

Reply:
Technically you could even use the current file in the second if clause
```go
	// Using `currentFile` here instead of underscore
	currentFile, ok := currentFilesMap[file.GetFileMeta().GetName()]

	// OMITTED other code for brevity
	
	// if file currently exists and file hash is different, file has been updated
	if ok && file.GetFileMeta().GetHash() != currentFile.GetFileMeta().GetHash() {
		file.Action = &updateAction
	}
```